Reformat HIP RTC (#3671)
ROCm-Developer-Tools/HIP: docs/how-to/hip_rtc.rst

                      0,                     // Number of headers
                      NULL,                  // Header sources
                      NULL);                 // Name of header files

:cpp:func:`hiprtcCreateProgram` API also allows you to add headers which can be
included in your RTC program. For online compilation, the compiler pre-defines
HIP device API functions, HIP specific types and macros for device compilation,
but doesn't include standard C/C++ headers by default. Users can only include
header files provided to :cpp:func:`hiprtcCreateProgram`.

After associating the kernel string with :cpp:struct:`hiprtcProgram`, you can
now compile this program using:

.. code-block:: cpp

  hiprtcCompileProgram(prog,     // hiprtcProgram
                      0,         // Number of options
                      options);  // Clang Options [Supported Clang Options](clang_options.md)

:cpp:func:`hiprtcCompileProgram` returns a status value which can be converted
to string via :cpp:func:`hiprtcGetErrorString`. If compilation is successful,
:cpp:func:`hiprtcCompileProgram` will return ``HIPRTC_SUCCESS``.

If the compilation fails, you can look up the logs via:

> 

Reviewer: It says "if the compilation fails", but it would be more precise if it says "if the compilation fails or produces warnings" i.e. it can be successful and still produce a log if there are warnings

Reviewer 2: @g-h-c Updated the text.

                                            0,                // Number of headers
                                            NULL,             // Header sources
                                            NULL);            // Name of header file

    if (rtc_ret_code != HIPRTC_SUCCESS) {
      std::cout << "Failed to create program" << std::endl;
      std::abort();
    }

    hipDeviceProp_t props;
    int device = 0;
    HIP_CHECK(hipGetDeviceProperties(&props, device));
    std::string sarg = std::string("--gpu-architecture=") +
        props.gcnArchName;  // device for which binary is to be generated

    const char* options[] = {sarg.c_str()};

    rtc_ret_code = hiprtcCompileProgram(prog,      // hiprtcProgram
                                        0,         // Number of options
                                        options);  // Clang Options
    if (rtc_ret_code != HIPRTC_SUCCESS) {
      std::cout << "Failed to create program" << std::endl;
      std::abort();
    }

    size_t logSize;
    HIPRTC_CHECK(hiprtcGetProgramLogSize(prog, &logSize));

    if (logSize) {
      std::string log(logSize, '\0');
      HIPRTC_CHECK(hiprtcGetProgramLog(prog, &log[0]));

>       std::cout << "Compilation failed with: " << log << std::endl;

Reviewer: I would say "Compilation failed or produced warnings"

Reviewer 2: Fixed.

  If the runtime fails to parse the defined string, or the environment variable
  is set to an empty string (""), the cache is disabled.

.. note::

  This cache is also shared with the OpenCL runtime shipped with ROCm.

HIPRTC specific options
===============================================================================

HIPRTC provides a few HIPRTC specific flags:

* ``--gpu-architecture`` : This flag can guide the code object generation for a
  specific GPU architecture. Example:
  ``--gpu-architecture=gfx906:sramecc+:xnack-``, its equivalent to
  ``--offload-arch``.

  * This option is compulsory if compilation is done on a system without AMD
    GPUs supported by HIP runtime.

  * Otherwise, HIPRTC will load the hip runtime and gather the current device
    and its architecture info and use it as option.

* ``-fgpu-rdc`` : This flag when provided during the
  :cpp:func:`hiprtcCreateProgram` generates the bitcode (HIPRTC doesn't convert
  this bitcode into ISA and binary). This bitcode can later be fetched using
  :cpp:func:`hiprtcGetBitcode` and :cpp:func:`hiprtcGetBitcodeSize` APIs.

Bitcode
-------------------------------------------------------------------------------

> In the usual scenario, the kernel associated with :cpp:struct:`hiprtcProgram` is

Reviewer: Typo:
*usual*

.. meta::
  :description: HIP runtime compiler (RTC)
  :keywords: AMD, ROCm, HIP, CUDA, RTC, HIP runtime compiler

.. _hip_runtime_compiler_how-to:

*******************************************************************************
Programming for HIP runtime compiler (RTC)
*******************************************************************************

HIP supports the kernels compilation at runtime with the ``hiprtc*`` APIs.
Kernels can be stored as a text string and can be passed to HIPRTC APIs
alongside options to guide the compilation.

.. note::

  * This library can be used on systems without HIP installed nor AMD GPU driver
    installed at all (offline compilation). Therefore, it doesn't depend on any
    HIP runtime library.

> 

Reviewer 3: ```suggestion
```

  * This library depends on Code Object Manager (comgr). You can try to
    statically link comgr into HIPRTC to avoid ambiguity.

> 

Reviewer 3: ```suggestion
```

  * Developers can bundle this library with their application.

Compilation APIs
===============================================================================

> To use HIPRTC functionality, HIPRTC header needs to be included first.

Reviewer 3: ```suggestion
To use HIPRTC functionality the header needs to be included: 
```

> :code:`#include <hip/hiprtc.h>`

Reviewer 3: ```suggestion

:code:`#include <hip/hiprtc.h>`
```

Reviewer 2: Fixed.

Kernels can be stored in a string:

.. code-block:: cpp

  static constexpr auto kernel_source {
  R"(
      extern "C"
      __global__ void vector_add(float* output, float* input1, float* input2, size_t size) {
        int i = threadIdx.x;
        if (i < size) {
          output[i] = input1[i] + input2[i];
        }
      }
  )"};

> Now to compile this kernel, it needs to be associated with

Reviewer 3: ```suggestion
To compile this kernel, it needs to be associated with
```

to string via :cpp:func:`hiprtcGetErrorString`. If compilation is successful,
:cpp:func:`hiprtcCompileProgram` will return ``HIPRTC_SUCCESS``.

If the compilation fails, you can look up the logs via:

.. code-block:: cpp

  size_t logSize;
  hiprtcGetProgramLogSize(prog, &logSize);

  if (logSize) {
    string log(logSize, '\0');
    hiprtcGetProgramLog(prog, &log[0]);
    // Corrective action with logs
  }

If the compilation is successful, you can load the compiled binary in a local
variable.

.. code-block:: cpp

  size_t codeSize;
  hiprtcGetCodeSize(prog, &codeSize);

  vector<char> kernel_binary(codeSize);
  hiprtcGetCode(prog, kernel_binary.data());

After loading the binary, :cpp:struct:`hiprtcProgram` can be destroyed.
:code:`hiprtcDestroyProgram(&prog);`

The binary present in ``kernel_binary`` can now be loaded via
:cpp:func:`hipModuleLoadData` API.

.. code-block:: cpp

  hipModule_t module;
  hipFunction_t kernel;

  hipModuleLoadData(&module, kernel_binary.data());
  hipModuleGetFunction(&kernel, module, "vector_add");

And now this kernel can be launched via ``hipModule`` APIs.

The full example is below:

.. code-block:: cpp
    

>   #include <hip/hip_runtime.h>

Reviewer 3: Does it need the hip_runtime.h header to be included?

Reviewer 2: We need to keep this include.

                  << std::endl;
        std::abort();
      }
    }
    std::cout << "Passed" << std::endl;

    HIP_CHECK(hipFree(dinput1));
    HIP_CHECK(hipFree(dinput2));
    HIP_CHECK(hipFree(doutput));
  }

Kernel Compilation Cache
===============================================================================

HIPRTC incorporates a cache to avoid recompiling kernels between program
executions. The contents of the cache include the kernel source code (including
the contents of any ``#include`` headers), the compilation flags, and the
compiler version. After a ROCm version update, the kernels are progressively
recompiled, and the new results are cached. When the cache is disabled, each
kernel is recompiled every time it is requested.

Use the following environment variables to manage the cache status as enabled or
disabled, the location for storing the cache contents, and the cache eviction
policy:

* ``AMD_COMGR_CACHE`` By default this variable has a value of ``0`` and the
  compilation cache feature is disabled. To enable the feature set the
  environment variable to a value of ``1`` (or any value other than ``0``). This

>   behavior may change in a future release.

Reviewer 3: ```suggestion
* ``AMD_COMGR_CACHE`` By default this variable has a value of ``0`` and the
  compilation cache feature is disabled. To enable the feature set the
  environment variable to a value of ``1`` (or any value other than ``0``). 
```

  is set to an empty string (""), the cache is disabled.

.. note::

  This cache is also shared with the OpenCL runtime shipped with ROCm.

HIPRTC specific options
===============================================================================

HIPRTC provides a few HIPRTC specific flags:

* ``--gpu-architecture`` : This flag can guide the code object generation for a
  specific GPU architecture. Example:
  ``--gpu-architecture=gfx906:sramecc+:xnack-``, its equivalent to
  ``--offload-arch``.

  * This option is compulsory if compilation is done on a system without AMD
    GPUs supported by HIP runtime.

  * Otherwise, HIPRTC will load the hip runtime and gather the current device
    and its architecture info and use it as option.

* ``-fgpu-rdc`` : This flag when provided during the
  :cpp:func:`hiprtcCreateProgram` generates the bitcode (HIPRTC doesn't convert
  this bitcode into ISA and binary). This bitcode can later be fetched using 
  :cpp:func:`hiprtcGetBitcode` and :cpp:func:`hiprtcGetBitcodeSize` APIs.

Bitcode
-------------------------------------------------------------------------------

In the usual scenario, the kernel associated with :cpp:struct:`hiprtcProgram` is

> compiled into the binary which can be loaded and run. However, if ``-fpu-rdc``

Reviewer 3: ```suggestion
compiled into the binary which can be loaded and run. However, if ``-fgpu-rdc``
```

Reviewer 2: Nice catch.

gfx10+ support execution of wavefront in CU mode and work-group processor mode
(WGP). Please refer to section 2.3 of `RDNA3 ISA reference <https://www.amd.com/content/dam/amd/en/documents/radeon-tech-docs/instruction-set-architectures/rdna3-shader-instruction-set-architecture-feb-2023_0.pdf>`_.

gfx9 and below only supports CU mode.

In WGP mode, 4 warps of a block can simultaneously be executed on the workgroup
processor, where as in CU mode only 2 warps of a block can simultaneously
execute on a CU. In theory, WGP mode might help with occupancy and increase the
performance of certain HIP programs (if not bound to inter warp communication),
but might incur performance penalty on other HIP programs which rely on atomics
and inter warp communication. This also has effect of how the LDS is split
between warps, please refer to `RDNA3 ISA reference <https://www.amd.com/content/dam/amd/en/documents/radeon-tech-docs/instruction-set-architectures/rdna3-shader-instruction-set-architecture-feb-2023_0.pdf>`_ for more information.

HIPRTC assumes **WGP mode by default** for gfx10+. This can be overridden by
passing ``-mcumode`` to HIPRTC compile options in

> :cpp:func:`hiprtcCompileProgram`.

Reviewer 3: ```suggestion
.. note::

HIPRTC assumes **WGP mode by default** for gfx10+. This can be overridden by
passing ``-mcumode`` to HIPRTC compile options in
:cpp:func:`hiprtcCompileProgram`.
```

Reviewer 2: Fixed.

IR assembly code. However, ``HIPRTC_JIT_INPUT_LLVM_BUNDLED_BITCODE`` and
``HIPRTC_JIT_INPUT_LLVM_ARCHIVES_OF_BUNDLED_BITCODE`` are only for bundled
bitcode and archive of bundled bitcode.

.. code-block:: cpp

  HIPRTC_JIT_INPUT_CUBIN = 0,
  HIPRTC_JIT_INPUT_PTX,
  HIPRTC_JIT_INPUT_FATBINARY,
  HIPRTC_JIT_INPUT_OBJECT,
  HIPRTC_JIT_INPUT_LIBRARY,
  HIPRTC_JIT_INPUT_NVVM,
  HIPRTC_JIT_NUM_LEGACY_INPUT_TYPES,
  HIPRTC_JIT_INPUT_LLVM_BITCODE = 100,
  HIPRTC_JIT_INPUT_LLVM_BUNDLED_BITCODE = 101,
  HIPRTC_JIT_INPUT_LLVM_ARCHIVES_OF_BUNDLED_BITCODE = 102,
  HIPRTC_JIT_NUM_INPUT_TYPES = (HIPRTC_JIT_NUM_LEGACY_INPUT_TYPES + 3)

Backward Compatibility of LLVM Bitcode/IR
-------------------------------------------------------------------------------

For HIP applications utilizing HIPRTC to compile LLVM bitcode/IR, compatibility
is assured only when the ROCm or HIP SDK version used for generating the LLVM
bitcode/IR matches the version used during the runtime compilation. When an
application requires the ingestion of bitcode/IR not derived from the currently
installed AMD compiler, it must run with HIPRTC and comgr dynamic libraries that
are compatible with the version of the bitcode/IR.

> comgr, a shared library, incorporates the LLVM/Clang compiler that HIPRTC relies

Reviewer 3: ```suggestion
`Comgr <https://github.com/ROCm/llvm-project/tree/amd-staging/amd/comgr>`_ is a shared library that incorporates the LLVM/Clang compiler that HIPRTC relies
```

-------------------------------------------------------------------------------

* ``HIPRTC_JIT_IR_TO_ISA_OPT_EXT`` - AMD Only. Options to be passed on to link
  step of compiler by :cpp:func:`hiprtcLinkCreate`.

* ``HIPRTC_JIT_IR_TO_ISA_OPT_COUNT_EXT`` - AMD Only. Count of options passed on
  to link step of compiler.

Example:

.. code-block:: cpp

  const char* isaopts[] = {"-mllvm", "-inline-threshold=1", "-mllvm", "-inlinehint-threshold=1"};
  std::vector<hiprtcJIT_option> jit_options = {HIPRTC_JIT_IR_TO_ISA_OPT_EXT,
                                              HIPRTC_JIT_IR_TO_ISA_OPT_COUNT_EXT};
  size_t isaoptssize = 4;
  const void* lopts[] = {(void*)isaopts, (void*)(isaoptssize)};
  hiprtcLinkState linkstate;
  hiprtcLinkCreate(2, jit_options.data(), (void**)lopts, &linkstate);

Error Handling
===============================================================================

HIPRTC defines the ``hiprtcResult`` enumeration type and a function
:cpp:func:`hiprtcGetErrorString` for API call error handling. ``hiprtcResult``
``enum`` defines the API result codes. HIPRTC APIs return ``hiprtcResult`` to
indicate the call result. :cpp:func:`hiprtcGetErrorString` function returns a

> string describing the given ``hiprtcResult`` code, e.g., HIPRTC_SUCCESS to 

Reviewer 3: ```suggestion
string describing the given ``hiprtcResult`` code, for example HIPRTC_SUCCESS to 
```

"HIPRTC_SUCCESS". For unrecognized enumeration values, it returns
"Invalid HIPRTC error code".

``hiprtcResult`` ``enum`` supported values and the
:cpp:func:`hiprtcGetErrorString` usage are mentioned below.

.. code-block:: cpp

  HIPRTC_SUCCESS = 0,
  HIPRTC_ERROR_OUT_OF_MEMORY = 1,
  HIPRTC_ERROR_PROGRAM_CREATION_FAILURE = 2,
  HIPRTC_ERROR_INVALID_INPUT = 3,
  HIPRTC_ERROR_INVALID_PROGRAM = 4,
  HIPRTC_ERROR_INVALID_OPTION = 5,
  HIPRTC_ERROR_COMPILATION = 6,
  HIPRTC_ERROR_LINKING = 7,
  HIPRTC_ERROR_BUILTIN_OPERATION_FAILURE = 8,
  HIPRTC_ERROR_NO_NAME_EXPRESSIONS_AFTER_COMPILATION = 9,
  HIPRTC_ERROR_NO_LOWERED_NAMES_BEFORE_COMPILATION = 10,
  HIPRTC_ERROR_NAME_EXPRESSION_NOT_VALID = 11,
  HIPRTC_ERROR_INTERNAL_ERROR = 12

.. code-block:: cpp

  hiprtcResult result;
  result = hiprtcCompileProgram(prog, 1, opts);
  if (result != HIPRTC_SUCCESS) {
  std::cout << "hiprtcCompileProgram fails with error " << hiprtcGetErrorString(result);
  }

HIPRTC General APIs
===============================================================================

> HIPRTC provides the following API for querying the version.
> 
> ``hiprtcVersion(int* major, int* minor)`` - This sets the output parameters
> major and minor with the HIP Runtime compilation major version and minor version
> number respectively.
> 

Reviewer 3: ```suggestion
HIPRTC provides ``hiprtcVersion(int* major, int* minor)`` for querying the version.
This sets the output parameters major and minor with the HIP Runtime compilation major version and minor version number respectively.

```

> Currently, it returns hardcoded value. This should be implemented to return HIP

Reviewer 3: ```suggestion
Currently, it returns hardcoded values. This should be implemented to return HIP
```

Kernel containing various definitions ``__global__`` functions/function
templates and ``__device__/__constant__`` variables can be stored in a string.

.. code-block:: cpp
    
  static constexpr const char gpu_program[] {
  R"(
  __device__ int V1; // set from host code
  static __global__ void f1(int *result) { *result = V1 + 10; }
  namespace N1 {
  namespace N2 {
  __constant__ int V2; // set from host code
  __global__ void f2(int *result) { *result = V2 + 20; }
  }
  }
  template<typename T>
  __global__ void f3(int *result) { *result = sizeof(T); }
  )"};

:cpp:func:`hiprtcAddNameExpression` is called with various name expressions
referring to the address of ``__global__`` functions and
``__device__/__constant__`` variables.

.. code-block:: cpp

  kernel_name_vec.push_back("&f1");
  kernel_name_vec.push_back("N1::N2::f2");
  kernel_name_vec.push_back("f3<int>");
  for (auto&& x : kernel_name_vec) hiprtcAddNameExpression(prog, x.c_str());
  variable_name_vec.push_back("&V1");
  variable_name_vec.push_back("&N1::N2::V2");
  for (auto&& x : variable_name_vec) hiprtcAddNameExpression(prog, x.c_str());

> After which, the program is compiled using :cpp:func:`hiprtcCompileProgram` and
> the generated binary is loaded using :cpp:func:`hipModuleLoadData`. And the
> mangled names can be fetched using :cpp:func:`hirtcGetLoweredName`.

Reviewer 3: ```suggestion
After which, the program is compiled using :cpp:func:`hiprtcCompileProgram`, the generated binary is loaded using :cpp:func:`hipModuleLoadData`, and the
mangled names can be fetched using :cpp:func:`hirtcGetLoweredName`.
```

.. code-block:: cpp

  for (decltype(variable_name_vec.size()) i = 0; i != variable_name_vec.size(); ++i) {
    const char* name;
    hiprtcGetLoweredName(prog, variable_name_vec[i].c_str(), &name);
  }

.. code-block:: cpp

  for (decltype(kernel_name_vec.size()) i = 0; i != kernel_name_vec.size(); ++i) {
    const char* name;
    hiprtcGetLoweredName(prog, kernel_name_vec[i].c_str(), &name);
  }

The mangled name of the variables are used to look up the variable in the module
and update its value.

.. code-block:: cpp

  hipDeviceptr_t variable_addr;
  size_t bytes{};
  hipModuleGetGlobal(&variable_addr, &bytes, module, name);
  hipMemcpyHtoD(variable_addr, &initial_value, sizeof(initial_value));

Finally, the mangled name of the kernel is used to launch it using the
``hipModule`` APIs.

.. code-block:: cpp

  hipFunction_t kernel;
  hipModuleGetFunction(&kernel, module, name);
  hipModuleLaunchKernel(kernel, 1, 1, 1, 1, 1, 1, 0, nullptr, nullptr, config);

> Please have a look at ``hiprtcGetLoweredName.cpp`` for the detailed example.

Reviewer 3: I found a reference for this at https://github.com/ROCm/HIP-Common/blob/develop/tests/src/hiprtc/hiprtcGetLoweredName.cpp, but i am not sure this is the one you want to point to. In any case, I think some kind of reference here would be helpful.

Versioning
===============================================================================

> HIPRTC follows the below versioning.

Reviewer 3: ```suggestion
HIPRTC uses the following versioning:
```

``library_types.h``, ``hip_math_constants.h``, ``hip_complex.h``,
``math_functions.h``, ``surface_types.h`` etc. from 6.1. HIPRTC users need not
include any HIP macros or constants explicitly in their header files. All of
these should get included via HIPRTC builtins when the app links to HIPRTC
library.

Deprecation notice
===============================================================================

* Currently HIPRTC APIs are separated from HIP APIs and HIPRTC is available as a
  separate library ``libhiprtc.so``/ ``libhiprtc.dll``. But on Linux, HIPRTC
  symbols are also present in ``libamdhip64.so`` in order to support the
  existing applications. Gradually, these symbols will be removed from HIP
  library and applications using HIPRTC will be required to explicitly link to
  HIPRTC library. However, on Windows ``hiprtc.dll`` must be used as the
  ``amdhip64.dll`` doesn't contain the HIPRTC symbols.

* Data types such as ``uint32_t``, ``uint64_t``, ``int32_t``, ``int64_t``
  defined in std namespace in HIPRTC are deprecated earlier and are being
  removed from ROCm release 6.1 since these can conflict with the standard

>   C++ data types. These data types are now prefixed with ``__hip__``, e.g.

Reviewer 3: ```suggestion
  C++ data types. These data types are now prefixed with ``__hip__``, for example
```
